BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py

```diff
@@ -112,4 +112,123 @@ def do_post_build(self, thebuilder):
                                             "  %s - %s" % (case.attrib['name'], result.text))
                                         failure_count += 1
 
+            if thebuilder.env.GetValue("TOOL_CHAIN_TAG") == "GCC5":
+                self.gen_code_coverage_gcc(thebuilder)
+            elif thebuilder.env.GetValue("TOOL_CHAIN_TAG").startswith ("VS"):
+                self.gen_code_coverage_msvc(thebuilder)
+            else:
+                logging.info("Skipping code coverage. Currently, support GCC and MSVC compiler.")
+
         return failure_count
+
+    def gen_code_coverage_gcc(self, thebuilder):
+        logging.info("Generating UnitTest code coverage")
+
+        buildOutputBase = thebuilder.env.GetValue("BUILD_OUTPUT_BASE")
+        workspace = thebuilder.env.GetValue("WORKSPACE")
+
+        # Generate base code coverage for all source files
+        ret = RunCmd("lcov", f"--no-external --capture --initial --directory {buildOutputBase} --output-file {buildOutputBase}/cov-base.info --rc lcov_branch_coverage=1")
+        if(ret != 0):
+            logging.error("UnitTest Coverage: Failed to build initial coverage data.")
+            return 1
+
+        # Coverage data for tested files only
+        ret = RunCmd("lcov", f"--capture --directory {buildOutputBase}/ --output-file {buildOutputBase}/coverage-test.info --rc lcov_branch_coverage=1")

```

Comment:
there is really inconsistent sep usage.  
What about using `{os.path.join(buildOutputBase, 'coverage-test.info'}` and same with the directory parameter and the dozens of other usages lower in this file.